pydantic_ai_slim/pydantic_ai/toolsets/searchable.py

```diff
@@ -0,0 +1,128 @@
+import logging
+import re
+from collections.abc import Callable
+from dataclasses import dataclass, field, replace
+from typing import Any, TypedDict
+
+from pydantic import TypeAdapter
+from typing_extensions import Self
+
+from .._run_context import AgentDepsT, RunContext
+from ..tools import ToolDefinition
+from .abstract import AbstractToolset, SchemaValidatorProt, ToolsetTool
+
+_SEARCH_TOOL_NAME = 'load_tools'
+
+
+class _SearchToolArgs(TypedDict):
+    regex: str
+
+
+def _search_tool_def() -> ToolDefinition:
+    return ToolDefinition(
+        name=_SEARCH_TOOL_NAME,
+        description="""Search and load additional tools to make them available to the agent.
+
+DO call this to find and load more tools needed for a task.
+NEVER ask the user if you should try loading tools, just try.
+""",
+        parameters_json_schema={
+            'type': 'object',
+            'properties': {
+                'regex': {
+                    'type': 'string',
+                    'description': 'Regex pattern to search for relevant tools',
+                }
+            },
+            'required': ['regex'],
+        },
+    )
+
+
+def _search_tool_validator() -> SchemaValidatorProt:
+    return TypeAdapter(_SearchToolArgs).validator
+
+
+@dataclass
+class _SearchTool(ToolsetTool[AgentDepsT]):
+    """A tool that searches for more relevant tools from a SearchableToolSet."""
+
+    tool_def: ToolDefinition = field(default_factory=_search_tool_def)
+    args_validator: SchemaValidatorProt = field(default_factory=_search_tool_validator)
+
+
+@dataclass
+class SearchableToolset(AbstractToolset[AgentDepsT]):
+    """A toolset that implements tool search and deferred tool loading."""
+
+    toolset: AbstractToolset[AgentDepsT]
+    _active_tool_names: set[str] = field(default_factory=set)
+
+    @property
+    def id(self) -> str | None:
+        return None  # pragma: no cover
+
+    @property
+    def label(self) -> str:
+        return f'{self.__class__.__name__}({self.toolset.label})'  # pragma: no cover
+
+    async def get_tools(self, ctx: RunContext[AgentDepsT]) -> dict[str, ToolsetTool[AgentDepsT]]:
+        logging.debug("SearchableToolset.get_tools")
+        all_tools: dict[str, ToolsetTool[AgentDepsT]] = {}
+        all_tools[_SEARCH_TOOL_NAME] = _SearchTool(
+            toolset=self,
+            max_retries=1,
```

Comment:
We may want to increase this, to give the model a few chances to fix its regex, if it submitted an invalid one the first time